File: web-ts/src/logic/CompoundPredicate.ts
/**
 * CompoundPredicate - base for compound predicates with iterative constraint solving.
 * @JAVA_REF org.modelingvalue.nelumbo.logic.CompoundPredicate
 */

import { List, Map, Set } from 'immutable';
import type { AstElement } from '../AstElement';
import { Variable } from '../Variable';
import { Node } from '../Node';
import type { Functor } from '../patterns/Functor';
import { Predicate } from './Predicate';
import { InferResult } from '../InferResult';
import type { InferContext } from '../InferContext';

// Late-bound reference to NBoolean (set from NBoolean.ts to avoid circular import)
let _NBooleanTRUE: (() => Predicate) | null = null;
let _NBooleanFALSE: (() => Predicate) | null = null;

export function _setNBooleanRefs(getTRUE: () => Predicate, getFALSE: () => Predicate): void {
  _NBooleanTRUE = getTRUE;
  _NBooleanFALSE = getFALSE;
}

/**
 * CompoundPredicate - a predicate containing sub-predicates.
 * @JAVA_REF org.modelingvalue.nelumbo.logic.CompoundPredicate
 */
export class CompoundPredicate extends Predicate {
  constructor(functor: Functor, elements: List<AstElement>, ...args: unknown[]) {
    super(functor, elements, ...args);
  }

  protected static fromDataCompound(data: unknown[], declaration?: Node): CompoundPredicate {
    const pred = Object.create(CompoundPredicate.prototype) as CompoundPredicate;
    (pred as unknown as { _data: unknown[] })._data = data;
    (pred as unknown as { _declaration: Node })._declaration = declaration ?? pred;
    (pred as any)._binding = null;
    (pred as any)._hashCodeCached = false;
    (pred as any)._hashCode = 0;
    (pred as any)._nrOfUnbound = -1;
    return pred;
  }

  protected override struct(data: unknown[], declaration?: Node): CompoundPredicate {
    return CompoundPredicate.fromDataCompound(data, declaration ?? this.declaration());
  }

  /**
   * Resolve this compound predicate.
   * Uses iterative shallow → reduce → deep inference with NBoolean.TRUE/FALSE replacement.
   * @JAVA_REF org.modelingvalue.nelumbo.logic.CompoundPredicate#resolve(InferContext)
   */
  override resolve(context: InferContext): InferResult {
    type BindingKey = Map<Variable, unknown>;
    let now: Map<BindingKey, Predicate>;
    let next: Map<BindingKey, Predicate> = Map<BindingKey, Predicate>().set(this.getBinding() ?? Map(), this as Predicate);
    let facts = Set<Predicate>();
    let falsehoods = Set<Predicate>();
    let cycles = Set<Predicate>();
    let completeFacts = true;
    let completeFalsehoods = true;
    const deep = context;
    const shallow = deep.toShallow();
    const reduce = deep.toReduce();

    do {
      now = next;
      next = Map<BindingKey, Predicate>();

      for (const [binding, predicate] of now.entries()) {
        // Phase 1: Shallow inference
        let result = predicate.infer(shallow);
        if (result.hasStackOverflow()) {
          return result;
        }
        if (!result.unresolvable()) {
          for (const pred of result.allFacts()) {
            let b = pred.getBinding();
            if (b !== null && !b.isEmpty()) {
              b = binding.merge(b);
              next = next.set(b, predicate.setBinding(b).replacePredicate(pred, _NBooleanTRUE!() as Predicate));
            }
          }
          for (const pred of result.allFalsehoods()) {
            let b = pred.getBinding();
            if (b !== null && !b.isEmpty()) {
              b = binding.merge(b);
              next = next.set(b, predicate.setBinding(b).replacePredicate(pred, _NBooleanFALSE!() as Predicate));
            }
          }
          completeFacts = completeFacts && result.completeFacts();
          completeFalsehoods = completeFalsehoods && result.completeFalsehoods();
          cycles = cycles.union(result.cycles());
        }

        // Phase 2: Reduce inference
        result = predicate.infer(reduce);
        if (result.hasStackOverflow()) {
          return result;
        } else if (result.isFalseCC()) {
          falsehoods = falsehoods.add(this.setBinding(binding));
        } else if (result.isTrueCC()) {
          facts = facts.add(this.setBinding(binding));
        } else {
          // Phase 3: Deep inference
          const resultPredicate = result.predicateOf();
          if (resultPredicate !== null) {
            result = resultPredicate.infer(deep);
            if (result.hasStackOverflow()) {
              return result;
            }
            if (!result.unresolvable()) {
              for (const pred of result.allFacts()) {
                let b = pred.getBinding();
                if (b !== null && !b.isEmpty()) {
                  b = binding.merge(b);
                  next = next.set(b, resultPredicate.setBinding(b).replacePredicate(pred, _NBooleanTRUE!() as Predicate));
                }
              }
              for (const pred of result.allFalsehoods()) {
                let b = pred.getBinding();
                if (b !== null && !b.isEmpty()) {
                  b = binding.merge(b);
                  next = next.set(b, resultPredicate.setBinding(b).replacePredicate(pred, _NBooleanFALSE!() as Predicate));
                }
              }
              completeFacts = completeFacts && result.completeFacts();
              completeFalsehoods = completeFalsehoods && result.completeFalsehoods();
              cycles = cycles.union(result.cycles());
            }
          }
        }
      }
    } while (!next.isEmpty());

    // @JAVA_REF: if no results found and both complete, set both to incomplete
    if (facts.isEmpty() && completeFacts && falsehoods.isEmpty() && completeFalsehoods) {
      completeFacts = false;
      completeFalsehoods = false;
    }

    return InferResult.of(facts, completeFacts, falsehoods, completeFalsehoods, cycles);
  }
}

File: src/main/java/org/modelingvalue/nelumbo/logic/CompoundPredicate.java
//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
// (C) Copyright 2018-2026 Modeling Value Group B.V. (http://modelingvalue.org)                                        ~
//                                                                                                                     ~
// Licensed under the GNU Lesser General Public License v3.0 (the 'License'). You may not use this file except in      ~
// compliance with the License. You may obtain a copy of the License at: https://choosealicense.com/licenses/lgpl-3.0  ~
// Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on ~
// an 'AS IS' BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the  ~
// specific language governing permissions and limitations under the License.                                          ~
//                                                                                                                     ~
// Maintainers:                                                                                                        ~
//     Wim Bast, Tom Brus                                                                                              ~
//                                                                                                                     ~
// Contributors:                                                                                                       ~
//     Victor Lap                                                                                                      ~
//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

package org.modelingvalue.nelumbo.logic;

import java.io.Serial;

import org.modelingvalue.collections.Entry;
import org.modelingvalue.collections.List;
import org.modelingvalue.collections.Map;
import org.modelingvalue.collections.Set;
import org.modelingvalue.nelumbo.AstElement;
import org.modelingvalue.nelumbo.Node;
import org.modelingvalue.nelumbo.lang.Functor;
import org.modelingvalue.nelumbo.lang.Type;
import org.modelingvalue.nelumbo.lang.Variable;

public abstract class CompoundPredicate extends Predicate {
    @Serial
    private static final long serialVersionUID = -4926802375244295351L;

    protected CompoundPredicate(Functor functor, List<AstElement> elements, Node declaration, Object... predicates) {
        super(functor, elements, declaration, predicates);
    }

    protected CompoundPredicate(Type type, List<AstElement> elements, Node declaration, Object... predicates) {
        super(type, elements, declaration, predicates);
    }

    protected CompoundPredicate(Object[] args, Node functorOrType, List<AstElement> elements,
            CompoundPredicate declaration) {
        super(args, functorOrType, elements, declaration);
    }

    @Override
    public InferResult resolve(InferContext context) {
        Map<Map<Variable, Object>, Predicate> now, next = Map.of(Entry.of(getBinding(), this));
        Set<Predicate> facts = Set.of(), falsehoods = Set.of(), cycles = Set.of();
        boolean completeFacts = true, completeFalsehoods = true;
        InferContext deep = context.toDeep(); // Resolve variables shallow (bind)
        InferContext shallow = context.toShallow(); // Resolve variables deep (bind)
        InferContext reduce = context.toReduce(); // Do not resolve variables but perform logic and simplify (rewrite)
        do {
            now = next;
            next = Map.of();
            for (Entry<Map<Variable, Object>, Predicate> entry : now) {
                Map<Variable, Object> binding = entry.getKey();
                Predicate predicate = entry.getValue();
                InferContext resolve = shallow;
                InferResult result = predicate.infer(reduce);
                if (result.hasStackOverflow()) {
                    return result;
                } else if (result.isFalseCC()) {
                    falsehoods = falsehoods.add(setBinding(binding));
                } else if (result.isTrueCC()) {
                    facts = facts.add(setBinding(binding));
                } else {
                    predicate = result.predicate();
                    resolve = deep;
                }
                result = predicate.infer(resolve);
                if (result.hasStackOverflow()) {
                    return result;
                } else if (!result.unresolvable()) {
                    for (Predicate pred : result.allFacts()) {
                        Map<Variable, Object> b = pred.getBinding();
                        if (!b.isEmpty()) {
                            b = binding.putAll(b);
                            next = next.put(b, predicate.setBinding(b).replace(pred, NBoolean.TRUE));
                        }
                    }
                    for (Predicate pred : result.allFalsehoods()) {
                        Map<Variable, Object> b = pred.getBinding();
                        if (!b.isEmpty()) {
                            b = binding.putAll(b);
                            next = next.put(b, predicate.setBinding(b).replace(pred, NBoolean.FALSE));
                        }
                    }
                    completeFacts &= result.completeFacts();
                    completeFalsehoods &= result.completeFalsehoods();
                    cycles = cycles.addAll(result.cycles());
                }
            }
        } while (!next.isEmpty());
        if (facts.isEmpty() && completeFacts && falsehoods.isEmpty() && completeFalsehoods) {
            completeFacts = false;
            completeFalsehoods = false;
        }
        return InferResult.of(this, facts, completeFacts, falsehoods, completeFalsehoods, cycles);
    }

}
